src/core/cgroup.c

```diff
@@ -2769,6 +2781,36 @@ static void unit_add_to_cgroup_oom_queue(Unit *u) {
                 log_error_errno(r, "Failed to enable cgroup oom event source: %m");
 }
 
+static int unit_check_cgroup_events(Unit *u) {
+        int r, i;
+        char *values[2] = {};
+
+        assert(u);
+
+        r = cg_get_keyed_attribute(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, "cgroup.events", STRV_MAKE("populated", "frozen"), values);
+        /* "frozen" key exists only in kernels >v5.2 so assume thawed cgroup before that. */
+        if (r != ENXIO && r < 0)
+                return r;
+
+        /* The cgroup.events notifications can be merged together so act as we saw the given state for the
+         * first time. The functions we call to handle given state are idempotent, which makes them
+         * effectively remember the previous state. */
+        if (values[0] == NULL || streq(values[0], "1"))
```

Comment:
doesn't matter, but we usually check strings for != NULL by relying on C's downgrade-to-bool feature, i.e. maybe:

```c
if (!values[0] || streq(values[0], "1")
``
Hmm, if the "populated" field is missing, maybe we should not change the cgroup empty queue at all, as we have no knowledge, and the empty queue processing verifies empty state anyway when it is dispatched... i.e. hence maybe:

```c
if (values[0]) {
        if (streq(values[0], "1")
                 …
         else
              …
}
```